src/main/java/uk/gov/justice/digital/hmpps/keyworker/services/KeyworkerService.java

```diff
@@ -132,12 +136,59 @@ public String startAutoAllocation(String agencyId) {
     }
 
     @PreAuthorize("#oauth2.hasScope('write')")
-    public void allocate(KeyworkerAllocationDto keyworkerAllocation) {
+    @Transactional
+    public void allocate(@Valid KeyworkerAllocationDto keyworkerAllocation) {
+        Validate.notNull(keyworkerAllocation);
+        verifyAgencySupport(keyworkerAllocation.getAgencyId());
 
-        restTemplate.postForObject(
-                "/key-worker/allocate",
-                new HttpEntity<>(keyworkerAllocation, CONTENT_TYPE_APPLICATION_JSON),
-                Void.class);
+        doAllocate(keyworkerAllocation);
+    }
+
+    private void doAllocate(KeyworkerAllocationDto newAllocation) {
+
+        /* TODO - validation
+        final String agencyId = bookingService.getBookingAgency(bookingId);
+        if (!bookingService.isSystemUser()) {
+            bookingService.verifyBookingAccess(bookingId);
+        }
+        validateStaffId(bookingId, newAllocation.getStaffId());
+        ==  @Override
+    public void checkAvailableKeyworker(Long bookingId, Long staffId) {
+        final String sql = getQuery("CHECK_AVAILABLE_KEY_WORKER");
+        try {
+            jdbcTemplate.queryForObject(sql, createParams("bookingId", bookingId, "staffId", staffId), Long.class);
+        } catch (EmptyResultDataAccessException ex) {
+            throw new EntityNotFoundException(
+                    String.format("Key worker with id %d not available for offender %d", staffId, bookingId));
+        }
+ SELECT DISTINCT(SLR.SAC_STAFF_ID)
+  FROM STAFF_LOCATION_ROLES SLR
+    INNER JOIN STAFF_MEMBERS SM ON SM.STAFF_ID = SLR.SAC_STAFF_ID
+      AND SM.STATUS = 'ACTIVE'
+    INNER JOIN OFFENDER_BOOKINGS OB ON OB.AGY_LOC_ID = SLR.CAL_AGY_LOC_ID
+      AND OB.OFFENDER_BOOK_ID = :bookingId
+      AND OB.ACTIVE_FLAG = 'Y'
+  WHERE SLR.SAC_STAFF_ID = :staffId
+    AND SLR.POSITION = 'AO'
+    AND SLR.ROLE = 'KW'
+    AND TRUNC(SYSDATE) BETWEEN TRUNC(SLR.FROM_DATE) AND TRUNC(COALESCE(SLR.TO_DATE,SYSDATE))
+    }*/
+
+        // Remove current allocation if any
+        repository.deactivate(newAllocation.getOffenderNo(), newAllocation.getDeallocationReason(), LocalDateTime.now());
```

Comment:
I'd rather it wasn't the responsibility of the allocate method to deallocate a previous allocation. I think it should check if there is an existing active allocation for the offender and throw an exception if there is. Let application logic outside this deal with deallocation.